design/proposal-crossplane-v2.md

```diff
@@ -0,0 +1,948 @@
+# Crossplane v2
+
+* Owner: Nic Cope (@negz)
+* Reviewers: Crossplane Steering Committee
+* Status: Draft
+
+
+## Executive Summary
+
+This document proposes a new version of Crossplane that’s simpler and better
+suited to managing applications, not just the infrastructure they depend on.
+
+In Crossplane v2:
+
+* Most composite resources (XRs) are namespaced
+* All managed resources (MRs) are namespaced
+* XRs can compose any Kubernetes resource - not just MRs
+* Claims are removed
+
+Under this proposal Crossplane v2 would be backward compatible with Crossplane
```

Comment:
Suggestion: Namespaced MRs disable LateInitialize by default

LateInitialize is problematic because it makes unmanaged fields managed, for example disk size of auto scaling resources and availability zone of multi-az resources with failover capability.

It could still be kept opt-in if there are use-cases for LateInitialize, but default off makes sense. I believe LateInitialize is less useful now that it was initialially, as status.atProvider now is consistent and complete.

There might also be per-resource issues, like https://github.com/crossplane-contrib/provider-upjet-aws/issues/1631

Reply:
I really like this idea. 

One consequence on the provider side would be that we'd probably want to update the upjet code that generates cross-resource references from terraform examples to point to `status.atProvider` rather than `spec.forProvider`, but I think that would probably be a good change anyways, and not difficult.

Reply:
I'd love to drop late init completely but I think someone recently mentioned a compelling use case for it that I've now forgotten. Either way making it off by default makes sense to me - raised https://github.com/crossplane/crossplane/issues/6512 to track.

I might not mention it in this design though. We're getting pretty granular with some of these changes and I don't think this doc needs to capture them exhaustively - more capture the overarching direction for v2.